### PR TITLE
properly use cluster id and provider id

### PR DESCRIPTION
**Pull request by stevenctl**
fixes https://github.com/istio/istio/issues/22093 and is pre-work for #29026 

The main advantage is that we can actually associate the ServiceEntry registry with a cluster (no longer rely on `ClusterID == ""`). We eventually plan on [reading WorkloadEntry from multiple clusters](https://docs.google.com/document/d/1Q_JqW7v1SC2Crg5ped9Bf8fyauglB-YqHYk7hwSuWOk/edit). To do this, we need to be able to shard the WorkloadInstances/Endpoints by source cluster.

### CODE REVIEW

**Comment by nmittler** [comment]
@stevenctl I'm not very familiar with this logic, but it seems that the main problem we're trying to solve is to let ServiceEntry  have precedence over k8s services. Is that accurate? If so, could we just assign a priority to each service registry and make sure that if a service is found in ServiceEntry, that it's used?

**Comment by stevenctl** [comment]
Not exactly. Really what I want is to allow having multiple service entry registries, and their generated IstioEndpoints be tied to a cluster id. We will only read workload entry from the other clusters, but reusing the existing controller.

Also just to cleanup the hack of using `clusterID == ""` to check service registry type.

File: pilot/pkg/serviceregistry/aggregate/controller.go [hunk 1]
```diff
@@ -120,11 +119,7 @@ func (c *Controller) Services() ([]*model.Service, error) {
 		// Race condition: multiple threads may call Services, and multiple services
 		// may modify one of the service's cluster ID
 		clusterAddressesMutex.Lock()
-		if r.Cluster() == "" { // Should we instead check for registry name to be on safe side?
-			// If the service does not have a cluster ID (ServiceEntries, CloudFoundry, etc.)
-			// Do not bother checking for the cluster ID.
-			// DO NOT ASSIGN CLUSTER ID to non-k8s registries. This will prevent service entries with multiple
-			// VIPs or CIDR ranges in the address field
+		if r.Provider() == serviceregistry.External {
```
Comment (nmittler) [inline pilot/pkg/serviceregistry/aggregate/controller.go:122]: should this be `r.Provider() != serviceregistry.Kubernetes` to match the logic in the other function?

File: pilot/pkg/serviceregistry/aggregate/controller.go [hunk 2]
```diff
@@ -218,29 +213,22 @@ func nodeClusterID(node *model.Proxy) string {
...
-	if registryClusterID == string(serviceregistry.Kubernetes) {
-		registryClusterID = selfClusterID
-	}
-
-	// Kube registries can have WorkloadEntry instances
-	if registryClusterID == "" || nodeClusterID == "" {
+func skipSearchingRegistryForProxy(nodeClusterID string, r serviceregistry.Instance) bool {
+	// Always search non-kube (usually serviceentry) registry.
+	// Check every registry if cluster ID isn't specified.
+	if r.Provider() == serviceregistry.External || nodeClusterID == "" {
```
Comment (nmittler) [inline pilot/pkg/serviceregistry/aggregate/controller.go:219]: same comment here

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/aggregate/controller.go:219]: good

Comment (stevenctl) [reply pilot/pkg/serviceregistry/aggregate/controller.go]: It makes sense to check non-kube here for sure.

File: pilot/pkg/serviceregistry/aggregate/controller.go (hunk 1 above)
Comment (stevenctl) [reply pilot/pkg/serviceregistry/aggregate/controller.go]: Here in particular I'm not sure if we want to check that... things that fall under this conditional will _not_ have clusterVips merged. Not sure if that's intended for all non-kube registries or only serviceeentry.

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**